### PR TITLE
Support Flexible Scheduler

### PR DESCRIPTION
**Pull request by benji-cannot-code**
**GitHub Issue:** [Support more learning rate schedulers](https://github.com/NovaSky-AI/SkyRL/issues/140)

**Changes:** 
1. Changed scheduler for FSDP strategy from hard-coded `constant_with_warmup` to flexible HF scheduler. 
2. Added scheduler to config
3. Removed trailing whitespaces in config and docs

**Testing:** 
1. Ran `uv run --isolated --extra dev pytest tests/cpu -q` sanity tests
2. vLLM smoke test
```
python - << 'PY'
from vllm import LLM, SamplingParams
m="TinyLlama/TinyLlama-1.1B-Chat-v1.0"
llm=LLM(model=m, tensor_parallel_size=1, trust_remote_code=True)
print(llm.generate(["hiiii im running test"], SamplingParams(max_tokens=16))[0].outputs[0].text.strip())
PY
```
3. E2E training test
```
uv run --isolated -m skyrl_train.entrypoints.main_base \
  --config examples/gsm8k/minimal.yaml \
  data.gsm8k_path=/skyrl-data/datasets/gsm8k \
  model.name="TinyLlama/TinyLlama-1.1B-Chat-v1.0" \
  generator.max_new_tokens=32 \
  trainer.steps=10 \
  trainer.global_batch_size=2 \
  generator.rollouts_per_iter=4 \
  ray.num_cpus=2 ray.num_gpus=1 \
  logging.output_dir=/skyrl-data/logs/gsm8k-smoke \
  checkpoint.dir=/skyrl-data/checkpoints/gsm8k-smoke
```

### CODE REVIEW

**Comment by benji-cannot-code** [comment]
My editor automatically trimmed trailing white spaces for the doc and the config. Is this fine?

File: skyrl-train/skyrl_train/distributed/fsdp_strategy.py [hunk 1]
```diff
@@ -276,9 +276,10 @@ def _fsdp_init_train_model(self, model, optimizer, scheduler):
...
             )
 
-            #  TODO(csy): add other schedulers, add more to config
-            actor_lr_scheduler = get_constant_schedule_with_warmup(
-                optimizer=actor_optimizer, num_warmup_steps=optim_config.num_warmup_steps
+            actor_lr_scheduler = get_scheduler(
+                optim_config.scheduler,
+                actor_optimizer,
+                num_warmup_steps=optim_config.num_warmup_steps,
             )
```
[the remark this answers is not in the document]
Comment (tyler-griggs) [reply skyrl-train/skyrl_train/distributed/fsdp_strategy.py]: This makes sense to me. We'll need to know the number of training steps for many schedulers. This information is known in the trainer, so it can/should be plumbed down.

File: skyrl-train/skyrl_train/workers/deepspeed/deepspeed_worker.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ def init_model(self, model_id_or_path):
         )
 
         actor_scheduler = get_scheduler(
-            "constant_with_warmup",
+            self.cfg.trainer.policy.optimizer_config.scheduler,
             actor_optim,
             num_warmup_steps=self.cfg.trainer.policy.optimizer_config.num_warmup_steps,
         )
```
[the remark this answers is not in the document]
Comment (tyler-griggs) [reply skyrl-train/skyrl_train/workers/deepspeed/deepspeed_worker.py]: Ditto here

File: skyrl-train/docs/configuration/config.rst [hunk 3]
```diff
@@ -175,19 +175,21 @@ For both the critic and policy model, we provide a common optimizer configuratio
...
        num_warmup_steps: 0
+       scheduler: "constant_with_warmup"
 
 - ``optimizer_config.lr``: Learning rate for the optimizer
 - ``optimizer_config.adam_betas``: Betas for AdamW optimizer.
 - ``optimizer_config.weight_decay``: L2 regularization strength for AdamW.
 - ``optimizer_config.max_grad_norm``: Gradient clipping parameter. The total L2 norm of the model gradients will be scaled to this value during training.
 - ``optimizer_config.offload_after_step``: Whether to offload optimizer state to CPU after step if colocated. When generation and training workers are colocated, we recommend using the default setting of ``true``. In some cases with non-colocation, it can be desirable to leave optimizer state on GPU memory to avoid offloading costs as well as additional CPU memory usage.
 - ``optimizer_config.num_warmup_steps``: Number of warmup steps for the learning rate scheduler.
+- ``optimizer_config.scheduler``: Which learning rate scheduler to use.
```
[the remark this answers is not in the document]
Comment (tyler-griggs) [reply skyrl-train/docs/configuration/config.rst]: Probably don't need to leave a link, but it would be good to mention that this config param is intended to align with schedulers in transformers.SchedulerType

File: skyrl-train/skyrl_train/workers/deepspeed/deepspeed_worker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (benji-cannot-code) [reply skyrl-train/skyrl_train/workers/deepspeed/deepspeed_worker.py]: Fixed

File: skyrl-train/skyrl_train/distributed/fsdp_strategy.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benji-cannot-code) [reply skyrl-train/skyrl_train/distributed/fsdp_strategy.py]: Fixed!

File: skyrl-train/docs/configuration/config.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (benji-cannot-code) [reply skyrl-train/docs/configuration/config.rst]: Done

**Comment by benji-cannot-code** [comment]
> Hey @benji-cannot-code , this is great, thank you!
> 
> Only two comments:
> 
> * Please plumb num_training_steps argument into `get_scheduler`, otherwise we can't actually use many of the schedulers.
> * Please test this with a new scheduler (ie, not `constant_with_warmup`). I don't believe we have a test to cover this, so you consider adding one or launching short training runs with ~2 new schedulers and validate that they are working.

Done. Tested both FSDP and DeepSpeed with linear and cosine (requires training steps).

**Comment by tyler-griggs** [comment]
Awesome, thanks @benji-cannot-code . Can you past a screenshot of the wandb log for the FSDP and DeepSpeed tests, especially for the learning rate plot? When we run mini training run tests, we typically share screenshots to keep track of what the curves look like.

**Comment by benji-cannot-code** [comment]
> Awesome, thanks @benji-cannot-code . Can you past a screenshot of the wandb log for the FSDP and DeepSpeed tests, especially for the learning rate plot? When we run mini training run tests, we typically share screenshots to keep track of what the curves look like.

Policy graphs (rl graphs match expectations): 

DeepSpeed with cosine scheduler:
Screenshot 2025-08-15 at 8 33 40 PM


FSDP with linear scheduler:
Screenshot 2025-08-15 at 8 35 32 PM

File: skyrl-train/skyrl_train/workers/deepspeed/deepspeed_worker.py [hunk 4]
```diff
@@ -265,6 +266,7 @@ def init_model(self, model_id_or_path):
             "constant_with_warmup",
```
[the remark this answers is not in the document]
Comment (tyler-griggs) [reply skyrl-train/skyrl_train/workers/deepspeed/deepspeed_worker.py]: @benji-cannot-code This is a good point ^

[the remark this answers is not in the document]
Comment (benji-cannot-code) [reply skyrl-train/skyrl_train/workers/deepspeed/deepspeed_worker.py]: Nice spot, fixed!

**Comment by tyler-griggs** [comment]
All looks good to me! Once you get a chance to confirm that checkpointing correctly saves-and-resumes the scheduler state (which it should), we can merge.

**Comment by benji-cannot-code** [comment]
> All looks good to me! Once you get a chance to confirm that checkpointing correctly saves-and-resumes the scheduler state (which it should), we can merge.

Thanks! Just did:

Ran FSDP with cosine scheduler and checkpoint every 5 steps. Killed the job after the first 5 steps (graph below) 
Screenshot 2025-08-18 at 5 59 16 PM
and restarted from there, and it looks fine (graph below). 
Screenshot 2025-08-18 at 6 03 16 PM
After the first pause, I paused again at the 35th step and resumed, this time using the same WandB run, and thus it's the same graph as above, and it also looks fine.
